kafka.go

```diff
@@ -0,0 +1,390 @@
+	topicDetail := &sarama.TopicDetail{}
+	topicDetail.NumPartitions = int32(1)
+	topicDetail.ReplicationFactor = int16(1)
+	topicDetail.ConfigEntries = make(map[string]*string)
+
+	topicDetails := make(map[string]*sarama.TopicDetail)
+	topicDetails[topicName] = topicDetail
+	request := sarama.CreateTopicsRequest{
+		Timeout:      time.Second * 15,
+		TopicDetails: topicDetails,
+	}
+
+	_, err := broker.CreateTopics(&request)
+	return err
+}
+
+// deleteTopic delete kafka topic
+func deleteTopic(broker *sarama.Broker, topicName string) error {
+	request := sarama.DeleteTopicsRequest{
+		Timeout: time.Second * 15,
+		Topics:  []string{topicName},
+	}
+
+	_, err := broker.DeleteTopics(&request)
+	return err
+}
+
+// unmarshal unmarshal received message into message struct
+func unmarshal(consumerMessage *sarama.ConsumerMessage) *Message {
+	var receivedMessage Message
+	err := json.Unmarshal(consumerMessage.Value, &receivedMessage)
+	if err != nil {
+		logrus.Error("Unable to unmarshal message from consumer in kafka : ", err)
+		return &Message{}
+	}
+	return &receivedMessage
+}
+
+// registerCallback add callback to map with topic and message Type as a key
+func registerCallback(topic, messageType string, callback func(message *Message, err error)) (isRegistered bool) {
+	if subscribeMap == nil {
+		subscribeMap = make(map[string]map[string]func(message *Message, err error))
+	}
+
+	if callbackMap, isTopic := subscribeMap[topic]; isTopic {
+		if _, isMsgType := callbackMap[messageType]; isMsgType {
+			return true
+		}
+	}
+
+	newCallbackMap := make(map[string]func(message *Message, err error))
+	newCallbackMap[messageType] = callback
+	subscribeMap[topic] = newCallbackMap
+	return false
+}
+
+// runCallback run callback function when receive a message
+func runCallback(receivedMessage *Message, consumerMessage *sarama.ConsumerMessage) {
+	callback := subscribeMap[consumerMessage.Topic][receivedMessage.MessageType]
+
+	if callback == nil {
+		logrus.Error(fmt.Sprintf("Callback not found for Topic : %s, Message Type : %s", consumerMessage.Topic,
+			receivedMessage.MessageType))
+		return
+	}
+
+	go callback(&Message{
+		Topic:       consumerMessage.Topic,
+		Message:     receivedMessage.Message,
+		MessageType: receivedMessage.MessageType,
+		Service:     receivedMessage.Service,
+		TraceId:     receivedMessage.TraceId,
+		MessageId:   receivedMessage.MessageId,
+	}, nil)
+}
+
+// processMessage process a message from kafka
+func processMessage(consumerMessage *sarama.ConsumerMessage) {
+	receivedMessage := unmarshal(consumerMessage)
+	runCallback(receivedMessage, consumerMessage)
+}
+
+// processResponseMessage process a message from kafka for publish with responses function
+func processResponseMessage(consumerMessage *sarama.ConsumerMessage, messageID string) bool {
+	receivedMessage := unmarshal(consumerMessage)
+
+	// check the request and response message ID should be equal
+	if messageID != receivedMessage.MessageId {
```

Comment:
is it possible that the receivedMessage is nil?

Reply:
I think its not possible because if unmarshal function get an error, it will return empty struct